text/trace/0234-payload-collection.md

```diff
@@ -0,0 +1,363 @@
+  Some example of payload data includes HTTP bodies, database queries (including items read or written), and messages produced or consumed from a queue.
+
+Currently, the support in OpenTelemetry for collecting such data is quite limited. We propose in this OTEP to add an API that will define a standard and extendable way for collecting payloads in traces.
+The implementation of this API is based on existing attributes, with proper semantic conventions.
+
+The possible value from collecting payloads is substantial. Using payload data, OpenTelemetry users can troubleshoot applications much more effectively in many cases - they can use that to understand exact data flows in their systems, reproduce problematic requests, or search for traces by specific payload information.
+Payload data could also be used to create ad-hoc analytics which helps monitoring the system.
+
+As many could argue, there are many challenges for effieciently collecting payloads from production applications in a generic way: handling sensitive data, performance implications, telemetry vendors support and others. Nevertheless, we think that this should not block OpenTelemetry users from manually collecting payload data already, whenever it fits their requirements.
+
+Having a standard semantic conventions for payload data can allow monitoring tools and others to use them OOTB (if they exist), without relying on custom attribute names.
+
+Ultimately, it could be beneficial if OpenTelemetry instrumentation SDKs will have OOTB optional payload collection as well, though this OTEP does NOT propose that, or takes into account that it will be added in the future.
+
+## Explanation
+
+While payload data can just be a binary buffer, most of the times it has a defined structure.
+For modern applications and APIs, the payload data is usually encoded as one of:
+
+- JSON
+- YAML
+- Protobuf
+- Avro
+- Plain strings
+- Blob
+
+Whenever the payload data has some meaningful structure (everything except a blob), that structure will be reflected in the semantic attribute of the relevant span. That way, processors downstream can easily access and manipulate the payload data.
+
+## Internal details
+
+### Semantic conventions
+
+We propose specifying a consistent naming for payload content attributes, and its
+related metadata. These conventions could help processors, backends and users to
+handle this kind of data.
+
+In this OTEP we only give ideas and recommendations for it, though
+a further discussion will be required for choosing the ideal naming when
+writing the actual specifications.
+
+The conventions we propose are added as a postfix to the base attribute name
+(for example, `http.request.body`):
+
+- **Data attribute**: `<attribute>.content`. Holds the decoded content of the payload data. Alternatives: `.data`, `.payload`.
+- **Size attribute**: `<attribute>.size`. Holds the number of bytes of the encoded payload data. Alternatives: `.length`, `.bytes`.
```

Comment:
I think it's important to keep the size as a separate attribute, since we may have an actual different content size (for example if we truncated part of it).